controller/rest/waf_rule.go

```diff
@@ -1242,3 +1283,451 @@ func importWaf(scope string, loginDomainRoles access.DomainRole, importTask shar
 
 	return nil
 }
+
+// caller has been verified for federal admin access right
+func replaceFedWafSensorGroups(WafSensors []*share.CLUSWafSensor, WafGroups []*share.CLUSWafGroup) bool {
+	wafSenMap := make(map[string]*share.CLUSWafSensor, len(WafSensors))
+	for _, wafsen := range WafSensors {
+		if wafsen.CfgType == share.FederalCfg {
+			wafSenMap[wafsen.Name] = wafsen
+		}
+	}
+	wafGrpMap := make(map[string]*share.CLUSWafGroup, len(WafGroups))
+	for _, wafgrp := range WafGroups {
+		if wafgrp.CfgType == share.FederalCfg {
+			wafGrpMap[wafgrp.Name] = wafgrp
+		}
+	}
+
+	lock, err := clusHelper.AcquireLock(share.CLUSLockPolicyKey, clusterLockWait)
+	if err != nil {
+		log.WithFields(log.Fields{"error": err}).Error("Failed to acquire cluster lock")
+		return false
+	}
+	defer clusHelper.ReleaseLock(lock)
+	var defsensor *share.CLUSWafSensor
+	defsensor = clusHelper.GetWafSensor(share.CLUSWafDefaultSensor)
+
+	/*
+	 * create default waf sensor.
+	 */
+	if defsensor == nil {
+		createDefaultWafSensor()
+		defsensor = clusHelper.GetWafSensor(share.CLUSWafDefaultSensor)
+		if defsensor == nil {
+			e := "sensor cannot be created in cluster!"
+			log.Error(e)
+			return false
+		}
+		log.Debug("Creating default waf sensor!")
+	}
+
+	if defsensor.RuleList == nil {
+		defsensor.RuleList = make(map[string]*share.CLUSWafRule)
+	}
+	if defsensor.PreRuleList == nil {
+		defsensor.PreRuleList = make(map[string][]*share.CLUSWafRule)
+	}
+
+	txn := cluster.Transact()
+	defer txn.Close()
+
+	exSensorMap := make(map[string]*share.CLUSWafSensor)
+	existingSensors := clusHelper.GetAllWafSensors()
+	for _, esen := range existingSensors {
+		if esen != nil {
+			exSensorMap[esen.Name] = esen
+		}
+		if esen != nil && esen.CfgType == share.FederalCfg {
+			if _, ok := wafSenMap[esen.Name]; !ok { // in existing but not in new. so delete it
+				for _, rn := range esen.RuleListNames {
+					_, foundInAll := defsensor.RuleList[rn]
+					if foundInAll {
+						delete(defsensor.RuleList, rn)
+					}
+				}
+				clusHelper.DeleteWafSensorTxn(txn, esen.Name)
+			}
+		}
+	}
+
+	for _, wsen := range wafSenMap {
+		if wsen != nil {
+			if wsen.Name == share.CLUSFedWafDefSyncSensor {
+				for rname, cdr := range wsen.RuleList {
+					cdr.ID = common.GetWafRuleID(defsensor)
```

Comment:
when federal rule is synced to managed cluster, the rule id is regenerated within managed cluster to avoid conflict, this id is only effective at the local cluster.